cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,261 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package throttling
+
+import (
+	"io"
+	"math"
+	"sync"
+	"time"
+)
+
+// tokenBucket is a simple token bucket used to refill credits for services
+// with increasing time.
+type tokenBucket struct {
+	creditsPerSecond float64
+	balance          float64
+	maxBalance       float64
+	lastTick         time.Time
+	timeNow          func() time.Time // For testing
+}
+
+func newTokenBucket(options CreditAccruerOptions, timeNow func() time.Time) *tokenBucket {
+	return &tokenBucket{
+		creditsPerSecond: options.CreditsPerSecond,
+		balance:          options.MaxBalance,
+		maxBalance:       options.MaxBalance,
+		timeNow:          timeNow,
+		lastTick:         timeNow(),
+	}
+}
+
+// Withdraw deducts as much of the total balance possible without exceeding
+// maxWithdrawal.
+func (t *tokenBucket) Withdraw(maxWithdrawal float64) float64 {
+	now := t.timeNow()
+	interval := now.Sub(t.lastTick)
+	t.lastTick = now
+	diff := interval.Seconds() * t.creditsPerSecond
+	t.balance = math.Min(t.balance+diff, t.maxBalance)
+	result := math.Min(t.balance, maxWithdrawal)
+	t.balance -= result
+	return result
+}
+
+// client represents a Jaeger-instrumented caller. Every Jaeger client
+// corresponds to an instrumented service, but it may be one of many instances
+// of that service. Therefore, there is a one-to-many relationship between
+// services and clients. The clients must request credits from the service's
+// creditAccruer and they all share the same underlying tokenBucket.
+// For example, if one client requests a withdrawal and receives all the service
+// credits, the next client withdrawal request will be rejected until the
+// credits refill.
+type client struct {
+	// id is a unique id for the instance of the Jaeger client
+	id string
+	// perOperationBalance maintains per-operation balances of withdrawn credits
+	perOperationBalance map[string]float64
+	// updateTime is the last time client was updated
+	updateTime time.Time
+}
+
+func newClient(id string, currentTime time.Time) *client {
+	return &client{
+		id:                  id,
+		updateTime:          currentTime,
+		perOperationBalance: map[string]float64{},
+	}
+}
+
+// Spend depletes the balance from the client for the given operation. The usage
+// would be when a client submits spans for operation X, it is spending those
+// credits it requested for operation X.
+func (c *client) Spend(operationName string, credits float64) bool {
+	balance := c.perOperationBalance[operationName]
+	if credits > balance {
+		return false
```

Comment:
this is unhappy path, but simply returning seems wrong. The client submitted the span, so it already spent the credits, we're just trying to record that fact. If the available balance is less than spent credits, then (a) we either have an error in the protocol of a dishonest client, so we need to report that somewhere, and (b) it's still better to reduce the balance to 0.

Reply:
Right. My concern here is what to do when our expectations are not met. This is where the C programmer in me would write `assert(credits <= balance)`, but Go clearly got rid of `assert` for a reason. In any case, I agree that setting the balance to zero makes more sense, so I'll do that. Also, will add some logging/reporting for the error.